sdk/communication/azure-communication-phonenumbers/src/main/java/com/azure/communication/phonenumbers/PhoneNumbersAsyncClient.java

```diff
@@ -144,9 +247,40 @@
      * @return A {@link PollerFlux} object.
      */
     @ServiceMethod(returns = ReturnType.LONG_RUNNING_OPERATION)
-    public PollerFlux<PollResult<Void>, Void> beginPurchasePhoneNumbers(String searchId) {
-        Objects.requireNonNull(searchId, "'searchId' can not be null.");
-        return null;
+    public PollerFlux<PhoneNumberOperation, Void> beginPurchasePhoneNumbers(String searchId) {
+        return beginPurchasePhoneNumbers(searchId, null);
+    }
+
+    PollerFlux<PhoneNumberOperation, Void> beginPurchasePhoneNumbers(String searchId, Context context) {
+        try {
+            Objects.requireNonNull(searchId, "'searchId' cannot be null.");
+            return new PollerFlux<>(defaultPollInterval,
+                purchaseNumbersInitOperation(searchId, context),
+                pollOperation(),
+                (pollingContext, firstResponse) -> Mono.error(logger.logExceptionAsError(new RuntimeException("Cancellation is not supported"))),
+                (pollingContext) -> Mono.empty());
+        } catch (RuntimeException ex) {
+            return PollerFlux.error(ex);
+        }
+    }
+
+    private Function<PollingContext<PhoneNumberOperation>, Mono<PhoneNumberOperation>> 
+        purchaseNumbersInitOperation(String searchId, Context context) {
+        return (pollingContext) -> {
+            return withContext(contextValue -> {
+                if (context != null) {
+                    contextValue = context;
+                }
+                return client.purchasePhoneNumbersWithResponseAsync(new PhoneNumberPurchaseRequest().setSearchId(searchId), contextValue)
+                .flatMap((PhoneNumbersPurchasePhoneNumbersResponse response) -> {
+                    pollingContext.setData("operationId", parseIdFromUrl(response.getDeserializedHeaders().getOperationLocation()));
+                    return client.getOperationAsync(pollingContext.getData("operationId"))
+                        .flatMap((PhoneNumberOperation result) -> {
+                            return Mono.just(result);
```

Comment:
Is there a reason you're using flatmap here

Reply:
That's a good point, I will update. thank you!